### PR TITLE
Enumerable methods feature

### PR DESCRIPTION
**Pull request by enelesmai**
Added custom new methods onto the existing Enumerable module:
* my_each
* my_each_with_index
* my_select
* my_all
* my_any
* my_none
* my_count
* my_map
* my_inject

Added main.rb file
Added my_test.rb file

### CODE REVIEW

File: .replit [hunk 1]
```diff
@@ -0,0 +1,2 @@
+language = "ruby"
+  run = "ruby main.rb"
```
Comment (suretrust) [inline .replit]: * [x] Files not related to the project should not be committed and added to .gitignore (e.g. `.repit`) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: my_enumerable_module.rb [hunk 2]
```diff
@@ -1,75 +1,84 @@
...
-        end
-        result
-    end
-    def my_none?
-        result = false
-        for i in 0...self.length
-            result = result || yield(self[i])
-        end
-        !result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_each.class != [1,2,3,4,5,6].each.class
```

File: my_enumerable_module.rb [hunk 3]
```diff
@@ -1,75 +1,84 @@
...
+
+  def my_each_with_index(start_index = 0)
+    for idx in 0...length
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_each_with_index.class != [1,2,3,4,5,6].each_with_index.class
```

File: my_enumerable_module.rb [hunk 4]
```diff
@@ -1,75 +1,84 @@
...
+  end
+
+  def my_select
+    result = []
+    my_each do |item|
+      result.push(item) if yield(item)
     end
-end
+    result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_select.class != [1,2,3,4,5,6].select.class
```

File: my_enumerable_module.rb [hunk 5]
```diff
@@ -1,75 +1,84 @@
...
-puts "\nmy_each"
-[1,2,3].each { |x| puts x }
-[1,2,3].my_each {|x| puts x}
+  def my_all?
+    result = true
+    my_each do |item|
+      result &&= yield(item)
+    end
+    result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_all?{ |num| num <= 6 } != [1,2,3,4,5,6].all?{ |num| num <= 6 }

[1,2,3,4,5,6].my_all?(Integer) != [1,2,3,4,5,6].all?(Integer)

%w[dog door rod blade].my_all?(/o/) != %w[dog door rod blade].all?(/o/)

[1,2,3,4,5,6].my_all?(3) != [1,2,3,4,5,6].all?(3)
```

File: my_enumerable_module.rb [hunk 6]
```diff
@@ -1,75 +1,84 @@
...
-#sets the start index
-[:foo, :bar, :baz].each.with_index(2) { |value, index| puts "#{index}: #{value}" }
-[:foo, :bar, :baz].my_each_with_index(6) { |value, index| puts "#{index}: #{value}" }
+  def my_any?
+    result = true
+    my_each do |item|
+      result ||= yield(item)
+    end
+    result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[nil, false, true, []].my_any? != [nil, false, true, []].any?

[1,2,3,4,5,6].my_any?(Integer) != [1,2,3,4,5,6].any?(Integer)

%w[dog door rod blade].my_any?(/d/) != %w[dog door rod blade].any?(/d/)

%w[dog door rod blade].my_any?('cat') != %w[dog door rod blade].any?('cat')
```

File: my_enumerable_module.rb [hunk 7]
```diff
@@ -1,75 +1,84 @@
...
-puts "\nmy_select"
-puts ["apple","berry","tomatoe","pinnepple","cocoa"].select { |x| x=="apple"}
-puts ["apple","berry","tomatoe","pinnepple","cocoa"].my_select { |x| x=="apple"}
+  def my_none?
+    result = false
+    my_each do |item|
+      result ||= yield(item)
+    end
+    !result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_none?(String) != [1,2,3,4,5].none?(String)

%w[dog door rod blade].my_none?(/z/) != %w[dog door rod blade].none?(/z/)

%w[dog door rod blade].my_none?(5) != %w[dog door rod blade].none?(5)
```

File: my_enumerable_module.rb [hunk 8]
```diff
@@ -1,75 +1,84 @@
...
+    return length unless block_given?
 
-puts "\nmy_any"
-puts %w[ant bear cat].any? { |word| word.length >= 4 } 
-puts %w[ant bear cat].my_any? { |word| word.length >= 4 } 
+    my_each do |item|
+      counter += 1 if yield item
+    end
+    counter
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_count(3) != [1,2,3,4,5].count(3)
```

File: my_enumerable_module.rb [hunk 9]
```diff
@@ -1,75 +1,84 @@
...
+      my_each do |item|
+        result.push(yield(item))
+      end
+    else
+      my_each do |item|
+        result.push(procedure.call(item))
+      end
+    end
+    result
+  end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_map.class != [1,2,3,4,5,6].map.class
```

File: my_enumerable_module.rb [hunk 10]
```diff
@@ -1,75 +1,84 @@
...
+    else
+      idx = 0
+      memo = initial
+    end
+    for i in idx...length
+      memo = yield(memo, self[i])
+    end
+    memo
+  end
+end
```
Comment (suretrust) [inline my_enumerable_module.rb]: * [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_inject(:+) != [1,2,3,4,5].inject(:+)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @enelesmai, 👋 

### Great job, you are definitely up to a great start! 🙌 Please see my comments on your code. 👇 

**Kindly submit another code review request once these issues have been fixed!** _Happy coding!_ 🎉

File: my_enumerable_module.rb (hunk 2 above)
Comment (enelesmai) [reply my_enumerable_module.rb]: @suretrust  Do you have any suggestions about this? I Can't find any documentation.

Comment (suretrust) [reply my_enumerable_module.rb]: You can use to_enum unless block is given.

Comment (enelesmai) [reply my_enumerable_module.rb]: got it! Thanks :)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @enelesmai     
#### Great Job so far, but there are a few more things required. 
- [x] Kindly refer to this doc for guidance https://ruby-doc.org/core-2.6.5/Enumerable.html 
- [x] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p ['car'].my_any?(Integer)` should return `false`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild','hellz'].my_any?(/z/)` should return `true`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p [2,32,].my_any?('cat')` should return `false`
- returns true if the block ever returns a value other than false or nil
`block =  proc { |num| num <= 2 }`
`p [3,43,4,234].my_any?(&block)` should return `false`
- [x] `my_all`  
- when no block or argument is given returns true when none of the collection members are false or nil:
`p [1, true, 'hi', []].my_all?` should return `true`
-  when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['d','d','d'].my_all?(/d/)` should return `true`
- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`
- Combines all elements of enum by applying a binary operation, specified by a block:
 `p (1..5).my_inject(4) { |prod, n| prod * n }` should return 480
- [x] `my_none?`:
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
`p ['hello','milk','milo'].my_none?(/i/)` should return `false`
- When no block or argument is given returns true only if none of the collection members is true
`p  [nil, false, nil, false] .my_none?` should return `true`











### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @enelesmai     
#### Great Job so far, you are almost there, just one more thing  required. 

- [x] `my_none?`:
- When no block or argument is given returns true only if none of the collection members is true
`p [nil, false, nil, true] .my_none?` should return `false`










### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by shshamim63** [comment]
## Status :tada:  Approved :tada: 
___
Hi @enelesmai :wave: , 
Awesome job :clap: . You have covered all the changes asked by the previous reviewers. The code looks neat and clean. But just found an issue, but as it was not mentioned previously by the previous reviewers, gonna mark it as optional. But believe me, adding some line of codes gonna make it perfect. 
___
### :clubs: Optional Changes :clubs: 
___
- `multiply_els` that uses `my_inject` as described on Odin ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
You forgot to add it to your source code. just add the lines after your module end. Cause multiply _els is not an enumerable method.
```ruby
def multiply_els(arr)
  p arr.my_inject(:*)
  p arr.my_inject { |memo, n| memo * n }
end
```
- Congrats you have added your sort of test cases. But believe me, this is not the way to write tests. You will add the actual test by using RSpec. So leaving those test cases now gonna be redundant. So, I will suggest you get rid of them now.
___
#### This :+1: is from me. You can go-ahead.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Twitter](https://twitter.com/Shshamim090)

**Comment by enelesmai** [comment]
Hello, @shshamim63 I already did what you suggested, thank you for that!

**Pull request merged**